internal/datasource/datasource.go

```diff
@@ -34,7 +34,7 @@ type Sourcer interface {
 		ui terminal.UI,
 		source *pb.Job_DataSource,
 		baseDir string,
-	) (string, func() error, error)
+	) (string, *pb.Job_DataSource_Ref, func() error, error)
```

Comment:
Probably don't need to do this here, but 4 return values is right on the verge of "this should be a data type" for me.

Reply:
Big agree. Its on my backburner to refactor the `internal/datasource` package since its scope is changing a bit anyways. Lets do that in another PR, after we get all this polling and stuff in.